com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/BreakpointManager.java

```diff
@@ -166,6 +171,28 @@ private void removeBreakpointsInternally(String source, IBreakpoint[] breakpoint
         return breakpointMap.values().toArray(new IBreakpoint[0]);
     }
 
+
+    /**
+     * Get the compiled expression map with breakpoint, it will be used in JdtEvaluationProvider#evaluateForBreakpoint for storing
+     * the compiled expression when the first time this conditional breakpoint is hit.
+     *
+     * @return the compiled expression map
+     */
+    public Map<IBreakpoint, Object> getBreakpointExpressionMap() {
```

Comment:
If the cache is only used by the evaluation Provider,  should it be better put insider the evaluation provider?

Reply:
the keys are managed by  breakpoint manager, and the values are managed by evaluation evaluation provider, I don't think its better for breakpoint manager to reference JdtEvaluationProvider

Reply:
JDT are storing expression in JavaLineBreakpoint@fCompiledExpressions

	/**
	 * The map of cached compiled expressions (ICompiledExpression) for this
	 * breakpoint, keyed by thread. This value must be cleared every time the
	 * breakpoint is added to a target.
	 */
	private Map<IJavaThread, ICompiledExpression> fCompiledExpressions = new HashMap<>();